### ISSUE TITLE
Using a partial type assumes partial knowledge

### ISSUE THREAD

**Issue by gvanrossum** [python/mypy#4088]
EDIT: the original example now works correctly, but this one still gives false-negative: https://github.com/python/mypy/issues/4088#issuecomment-335515687 

It seems when you use a variable with a partial type in an expression, the expression is type-checked with the knowledge gathered so far as the type of the variable. But this is not always correct. I found this on a more complex example involving an instance attribute. But here's a simpler repro:
```py
from typing import Optional

def f() -> None:
    a = None
    while True:
        g(a)  # <-- No error here!
        a = [1]

def g(a: Optional[str]) -> None:
    pass
```

**Comment by JukkaL** [comment]
Within a loop the type of a variable with a partial `None` type should be treated as unknown (unless it's defined within the loop, maybe). Mypy should require a type annotation for `a`. 

Note that partial list types work correctly:

```py
def f() -> None:
    a = []  # Need type annotation for variable
    while True:
        g(a)
        a.append(1)

def g(a: str) -> None:
    pass
```

**Comment by gvanrossum** [comment]
To whoever plans to fix this: it's not just loops. The actual example had something like
```py
def g(a: Optional[str]) -> None: ...
class C:
    a = None
    def foo(self) -> None:
        g(a)
    def bar(self) -> None:
        self.a = [1]
```

**Comment by ilevkivskyi** [comment]
https://github.com/python/mypy/issues/8158 gives another example.

**Comment by ilevkivskyi** [comment]
This now works correctly, I get:
```
testloop.py:6: error: Argument 1 to "g" has incompatible type "list[int] | None"; expected "str | None"  [arg-type]
```

**Comment by ilevkivskyi** [comment]
Oh well, the non-loop example still gives a false-negative, I will update description.

**Comment by JukkaL** [comment]
The non-loop example works as expected when using `--local-partial-types`, which we plan to enable by default in 2.0, so I think this can be closed.
